gpkit/nomials/map.py

```diff
@@ -0,0 +1,160 @@
+        out = NomialMap()
+        for exp in self:
+            if varkey in exp:
+                exp = HashVector(exp)
+                x = exp[varkey]
+                c = self[exp] * x
+                if x is 1:
+                    del exp[varkey]
+                else:
+                    exp[varkey] = x-1
+                out[exp] = c
+        out.set_units(self.units, 1.0/varkey.units if varkey.units else None)
+        return out
+
+    def sub(self, substitutions, varkeys, parsedsubs=False):
+        "Applies substitutions to a NomialMap"
+        # pylint: disable=too-many-locals, too-many-branches
+        if parsedsubs or not substitutions:
+            fixed = substitutions
+        else:
+            fixed = parse_subs(varkeys, substitutions)
+
+        if not substitutions:
+            if not self.expmap:
+                self.expmap, self.csmap = {exp: exp for exp in self}, {}
+            return self
+
+        cp = NomialMap()
+        cp.units = self.units
+        cp.expmap, cp.csmap = {}, {}
+        varlocs = defaultdict(set)
+        for exp, c in self.items():
+            new_exp = HashVector(exp)
+            cp.expmap[exp] = new_exp
+            cp[new_exp] = c
+            for vk in new_exp:
+                varlocs[vk].add((exp, new_exp))
+
+        for vk in varlocs:
+            if vk in fixed:
+                expval = []
+                exps, cval = varlocs[vk], fixed[vk]
+                if isinstance(cval, Strings):
+                    descr = dict(vk.descr)
+                    del descr["name"]
+                    cval = VarKey(name=cval, **descr)
+                if hasattr(cval, "hmap"):
+                    expval, = cval.hmap.keys()  # TODO: catch "can't-sub-posys"
+                    cval = cval.hmap  # pylint: disable=redefined-variable-type
+                if hasattr(cval, "to"):
+                    cval = mag(cval.to(vk.units or DIMLESS_QUANTITY))
+                    if isinstance(cval, NomialMap) and cval.keys() == [{}]:
+                        cval, = cval.values()
+                if expval:
+                    cval, = cval.values()
+                exps_covered = set()
+                for o_exp, exp in exps:
+                    x = exp[vk]
+                    powval = float(cval)**x if cval != 0 or x >= 0 else np.inf
+                    cp.csmap[o_exp] = powval * cp.csmap.get(o_exp, self[o_exp])
+                    if exp in cp and exp not in exps_covered:
+                        c = cp.pop(exp)
+                        del exp[vk]
+                        for key in expval:
+                            exp[key] = expval[key]*x + exp.get(key, 0)
+                        exp._hashvalue = None  # reset hash, # pylint: disable=protected-access
+                        cp[exp] = powval * c + cp.get(exp, 0)
+                        exps_covered.add(exp)
+                        cp.remove_zeros(just_monomials=True)  # pylint: disable=protected-access
+        return cp
```

Comment:
yup! This is replacing the much longer `substitution` code from `substitution.py`, which had even more special cases